### PR TITLE
Add expand-all action in search

### LINKED ISSUE #28090: Feature: Expand button in search
Hi all,

Currently when I search for a string and there are lots of results in lots of files, files with less than 10 results are automatically expanded. There is also a button "Collapse" that collapses all the files, but I can't seem to find a button to do the exact opposite - to expand all the files in search results. Could we get an "Expand" button please? It wouldn't even take up extra space - it could be shown instead of "Collapse" button when all files in search results are collapsed, and vice versa.

Thanks in advance, I use VSCode daily at work for half a year now and I'm really enjoying it!!

### PR DESCRIPTION
**Pull request by okmttdhr**
This PR fixes https://github.com/microsoft/vscode/issues/28090

[Image from Gyazo](https://gyazo.com/060e65b6ab90c22a6f1595a10c9d3ec4)

(Please let me know if there's a smart way to implement unit tests).

### CODE REVIEW

File: src/vs/workbench/contrib/search/browser/searchActions.ts [hunk 1]
```diff
@@ -363,6 +363,60 @@ export class CollapseDeepestExpandedLevelAction extends Action {
...
+export class ExpandAllAction extends Action {
+
+	static readonly ID: string = 'search.action.expandSearchResults';
+	static LABEL: string = nls.localize('ExpandAllAction.label', "Expand All");
+
+	constructor(id: string, label: string,
+		@IViewletService private readonly viewletService: IViewletService,
+		@IPanelService private readonly panelService: IPanelService
+	) {
+		super(id, label, 'search-action codicon-chrome-restore');
```
Comment (okmttdhr) [inline src/vs/workbench/contrib/search/browser/searchActions.ts]: How can I add a new `codicon` for `expand-all` ?
(`codicon-chrome-restore` is just a placeholder)

**Comment by JacksonKearl** [comment]
Thanks for taking this on! As discussed in the original issue, I think it would be better if this were only visible when there are items to expand. The trick is the three states:

- Everything collapsed => Show "Expand All"
- Everything expanded => Show "Collapse All"
- Otherwise => ???

As the default behavior for search is now to show all results expanded, I think we should lean towards providing Collapse functionality, and the "Expand All" action should only appear if everything is collapsed, with only the "Collapse All" being visible otherwise.

It might even make sense to only add a context menu entry for expanding all, as with #84087 the search results will almost always be fully expanded

cc @roblourens for ideas

**Comment by roblourens** [comment]
I agree, would rather not see two buttons at the same time. Just having a context menu option would be easier...

**Comment by okmttdhr** [comment]
Thank you for your feedback!

I'll try to hide/show each button according to items.

I'm thinking about adding property like `this.visible` to `ExpandAllAction` and `CollapseDeepestExpandedLevelAction`, but do you have any ideas??

**Comment by okmttdhr** [comment]
Hello @JacksonKearl (CC: @roblourens )
I updated this PR, could you please take a look at it?

**Comment by JacksonKearl** [comment]
Dec-12-2019 19-07-28

I'm noticing some unexpected behavior, where in a multiroot workspace if I expend the first node then run the action, everything expands. But if I expand any other node, the all collapse.

File: src/vs/workbench/contrib/search/browser/searchActions.ts (hunk 1 above)
Comment (JacksonKearl) [reply src/vs/workbench/contrib/search/browser/searchActions.ts]: @misolori Can we get `expand-all` codicon? Probably just `collapse-all` but with a plus.

**Review by JacksonKearl — APPROVED** [review]

**Comment by JacksonKearl** [comment]
I fixed the issue with the first node being treated differently (the while-loop was skipping the first node, changed to do-while) and added a hook to update the action when the tree state is toggled outside of this action. Should be good to go once we get an icon. Thanks for your work on this!

**Comment by misolori** [comment]
@roblourens @JacksonKearl I updated Codicons in master to now include a `expand-all` icon:

image

**Review by roblourens — APPROVED** [review]

**Comment by okmttdhr** [comment]
> I fixed the issue with the first node being treated differently (the while-loop was skipping the first node, changed to do-while) and added a hook to update the action when the tree state is toggled outside of this action. Should be good to go once we get an icon. Thanks for your work on this!

Thank you for your fixing and merging this PR 🙇✨☺️

**Pull request merged**